middleware/auth.js

```diff
@@ -0,0 +1,207 @@
+
+        if (correctApp) {
+            return data['user_id'];
+        } else {
+            return Promise.reject(new Error('Token from Google does not contain correct app information, or is invalid'));
+        }
+    })
+        .catch((error) => {
+            logger.log('info', error);
+            return Promise.reject(new Error('Something went wrong during the call to the Google token validation endpoint'));
+        });
+}
+
+/*
+
+ Helper Function
+ Used to validate a federated token of the supplied type
+
+ Params:
+    token - the token to authenticate
+    authType - the authentication that was used to get the token (facebook, google, etc.)
+
+ Returns:
+    Promise object, resolves to the user_id from within the provided token, or rejected if the token is invalid
+
+*/
+function getAuthDataFromIdp(token, authType) {
+    logger.log('info', 'getAuthDataFromIdp');
+
+    switch (authType) {
+        case 'facebook':
+            return validateFacebookToken(token);
+        case 'google':
+            return validateGoogleToken(token);
+    }
+}
+
+/*
+
+ Helper Function
+ Called by all secured APIs to validate that the supplied federated token is valid
+
+ Params:
+    token - the token to authenticate
+    authType - the authentication that was used to get the token (facebook, google, etc.)
+
+ Returns:
+    Promise object, resolves to the user_id from within the provided token, or rejected if the token is invalid
+
+*/
+async function validateFederatedToken(token, authType) {
+    logger.log('info', 'validateFederatedToken');
+
+    if (!token || !authType) {
+        return Promise.reject(new Error('AuthType and Token are required to validate token'));
+    }
+
+    if (!(SUPPORTED_AUTH_TYPES.includes(authType))) {
+        return Promise.reject(new Error('AuthType must be in the supported list of auth types'));
+    }
+
+    try {
+        const user_data = await getAuthDataFromIdp(token, authType);
+        logger.log('info', 'Data');
+        logger.log('info', user_data);
+        return user_data;
+    } catch (error) {
+        logger.log('info', error);
+        return Promise.reject(new Error('Generic error'));
+    }
+}
+
+async function validateIdentityToken(token) {
+    if (!token) {
+        return Promise.reject(new Error('No valid token found'));
+    }
+
+    const decoded = JSON.parse(Buffer.from(token, 'base64').toString());
+    logger.log('info', 'Decoded token');
+    logger.log('info', decoded);
+
+    if (!decoded || !decoded.fedToken) {
+        return Promise.reject(new Error('No valid token found'));
+    }
+
+    // TODO: Consider validating that the ID in the federated token matches the same user specified by the user_id in the ID token
+    return validateFederatedToken(decoded.fedToken, decoded.authType);
+}
+
+exports.authorize = function(req, res, next) {
+    if (req.path.includes('token/')) {
+        validateFederatedToken(req.query.fedToken, req.params.authType)
```

Comment:
It looks like the new route format causes the req.params.authType to be empty. The value of that variable is "undefined", which makes sense when I look at the serialized version of req (below). 

Maybe it makes sense to make both params to the API querystring params?

Note, this is just a subsection of the full 'req' serialization: 

`	"rawHeaders" : ["Accept", "application/json, text/plain, */*", "User-Agent", "axios/0.18.0", "Host", "localhost:3002", "Connection", "close"],
	"trailers" : {},
	"rawTrailers" : [],
	"aborted" : false,
	"upgrade" : false,
	"url" : "/token/facebook?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
	"method" : "GET",
	"statusCode" : null,
	"statusMessage" : null,
	"_consuming" : false,
	"_dumped" : false,
	"baseUrl" : "",
	"originalUrl" : "/token/facebook?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
	"_parsedUrl" : {
		"protocol" : null,
		"slashes" : null,
		"auth" : null,
		"host" : null,
		"port" : null,
		"hostname" : null,
		"hash" : null,
		"search" : "?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
		"query" : "fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
		"pathname" : "/token/facebook",
		"path" : "/token/facebook?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
		"href" : "/token/facebook?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD",
		"_raw" : "/token/facebook?fedToken=EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD"
	},
	"params" : {},
	"query" : {
		"fedToken" : "EAAXdPNyPDSYBAFQzESBUD1eZBR4e6yAJECyKB3uFAzTRAN6pUjvKZAQO9tJJCWT8DE38C6YViRRTGZAs5MgSEizrWPKn1LevNTfsc5rOQVWmuDUNqEkLlZAep5Q3dsLzkLdLVMXzE04wGQZCA8ZB4t92uZAjCkOzieGpYkzSf7hDxfcs6JVuDJEsDVfmyLyoUz147PU4ELuD4yxjfpml264VfvrUk8ZBW6BMSJvlDWE3HAZDZD"
	},`